### PR TITLE
[FLINK-6302] Documentation build error on ruby 2.4

### PR DESCRIPTION
**Pull request by mtunique**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by StephanEwen** [comment]
Would be great if someone else could confirm that this works. Ideally, we get a +1 from a Linux user, a MacOS X user, and a Cygwin user...

**Comment by haohui** [comment]
I hit the same problem on my Mac. Let me try it out later today.

**Comment by haohui** [comment]
+1 (non-binding)
Tested on OSX Sierra. The documentation builds successfully.

File: docs/Gemfile.lock [hunk 1]
```diff
@@ -30,18 +30,19 @@ GEM
       redcarpet (~> 3.1)
       safe_yaml (~> 1.0)
       toml (~> 0.1.0)
-    jekyll-coffeescript (1.0.1)
+    jekyll-coffeescript (1.0.2)
       coffee-script (~> 2.2)
+      coffee-script-source (~> 1.11.1)
     jekyll-gist (1.4.0)
-      octokit (~> 4.3.0)
+      octokit (~> 4.2)
```
Comment (zentol) [inline docs/Gemfile.lock]: Why are we downgrading a dependency?

File: docs/Gemfile.lock [hunk 2]
```diff
@@ -30,18 +30,19 @@ GEM
       redcarpet (~> 3.1)
       safe_yaml (~> 1.0)
       toml (~> 0.1.0)
-    jekyll-coffeescript (1.0.1)
+    jekyll-coffeescript (1.0.2)
       coffee-script (~> 2.2)
+      coffee-script-source (~> 1.11.1)
```
Comment (zentol) [inline docs/Gemfile.lock]: What's this about?

**Comment by zentol** [comment]
I can confirm that you can build the docs on W10 with the Linux Subsystem.

File: docs/Gemfile.lock (hunk 1 above)
Comment (mtunique) [inline docs/Gemfile.lock]: [https://github.com/jekyll/jekyll-gist/blob/master/jekyll-gist.gemspec](https://github.com/jekyll/jekyll-gist/blob/master/jekyll-gist.gemspec) In that file octokit version is 4.2.   I am sorry about I don't have much knowledge about ruby, I don't know why it's 4.2 in previous.

File: docs/Gemfile.lock (hunk 2 above)
Comment (mtunique) [inline docs/Gemfile.lock]: I find [https://github.com/jekyll/jekyll/pull/5662](https://github.com/jekyll/jekyll/pull/5662), that PR add the dependency.

Comment (zentol) [inline docs/Gemfile.lock]: For which jekyll version was that PR merged?

File: docs/Gemfile.lock (hunk 1 above)
Comment (zentol) [inline docs/Gemfile.lock]: Please revert this change; it worked before with 4.3.0 and this version is also referenced in other parts of the file.

Comment (StephanEwen) [inline docs/Gemfile.lock]: Does the downgrade fix an issue? If yes, its okay to do it, I think

**Comment by StephanEwen** [comment]
I am not familiar with the Ruby / Bundler / etc space, but it looks like `Gemfile` is the original source of truth and `Gemfile.lock` is in some way derived from that.
Is it correct to only update `Gemfile.lock` and not `Gemfile`?

@uce or @alpinegizmo can you maybe shed some light here?

File: docs/Gemfile.lock (hunk 1 above)
Comment (zentol) [inline docs/Gemfile.lock]: There was never an explicit version downgrade in `jekyll-gist`, they always had the `4.3.0` `octokit` dependency since it was added.

Similarly, since we added the `jekyll-gist`/`octokit` dependency the version of `octokit` was set to `4.3.0`.

In conclusion, it doesn't appear like it's related to a bug fix. We just always relied on a higher version than `jekyll-gist` did, and it worked out fine.

**Comment by zentol** [comment]
The `GemFile` should be updated as well. It would still work as long as the `GemFile.lock` file is present. If it deleted however the bundler would load the dependencies from `GemFile` and create a new `GemFile.lock`.

**Comment by alpinegizmo** [comment]
I prefer a conservative approach for as long as we still have to support building the docs on Ruby 1.9, so I'd keep the number of updated gems as small as possible.

As for adding support for Ruby 2.4, both docs/Gemfile and docs/ruby2/Gemfile should be updated, along with their corresponding (and derived) Gemfile.lock files. Moreover, the octokit downgrade is not needed. 

I'm not sure what motivates some of the proposed gem upgrades. I was able to get the docs to build on ruby 2.4 by making changing both Gemfiles as shown below. This leaves the classifier-reborn, ffi, libv8, posix-spawn, rb-inotify, redcarpet, and sass gems as they are rather than upgrading them. Do we need newer versions of any of those?

-gem 'json'
+gem 'json', '2.0.4'
+gem 'yajl-ruby', '1.2.2'
+gem 'jekyll-coffeescript', '1.0.2'

With these upgrades, the build script still works for me on Ruby 1.9.3 and 2.3.3, as well as 2.4.0.

**Comment by StephanEwen** [comment]
@mtunique What do you think about @alpinegizmo 's suggestion?

**Comment by mtunique** [comment]
@StephanEwen I think it is the best way to solve this issue. I will follow the suggestion.

**Comment by alpinegizmo** [comment]
This looks good, thank you. However, I now understand what was going on with the downgrade of the octokit dependency. I think we should first merge https://github.com/apache/flink/pull/4043 and then rebase this PR. @zentol @StephanEwen

**Comment by StephanEwen** [comment]
@alpinegizmo @zentol Now that #4043 is merged, can we rebase this one?

**Comment by alpinegizmo** [comment]
Yes, let's get this done.

**Comment by zentol** [comment]
merging.
